app/src/components/core/metaschema/metaSchema.ts

```diff
@@ -16,102 +16,146 @@ module app.core.metaschema {
+                    elements = schema['allOf'][index]['properties']['elements'];
+                }
+            } else if (schema.hasOwnProperty('properties') && schema['properties'].hasOwnProperty('elements')) {
+                elements = schema['properties']['elements'];
+            }
 
-                if (elements.hasOwnProperty('type') && elements['type'] === 'array' && elements.hasOwnProperty('items')) {
-                    var items = elements['items'];
-                    if (items.hasOwnProperty('type')) {
-                        if (items['type'] === 'array') {
-                            if (!items.hasOwnProperty('$ref')) {
-                                // a new object is introduced, otherwise $ref would be used
+            if (elements && elements.hasOwnProperty('type') && elements['type'] === 'array' && elements.hasOwnProperty('items')) {
+                var items = elements['items'];
+                if (items.hasOwnProperty('type')) {
+                    if (items['type'] === 'array') {
+                        if (!items.hasOwnProperty('$ref')) {
+                            // a new object is introduced, otherwise $ref would be used
 
-                            }
-                        } else if (items['type'] === 'object') {
-                            // we are sure that there is a new object introduced here
-                            if (items.hasOwnProperty('properties') && items['properties'].hasOwnProperty('type')) {
-                                if (items['properties']['type'].hasOwnProperty('enum')) {
-                                    _.forEach(items['properties']['type']['enum'], (name:string) => {
-                                        result.push(new Definition(name.toLowerCase(), Metaschema.cleanupDataschema(items), Metaschema.retrieveAcceptedElements(items, rootElements)));
-                                    });
-                                }
+                        }
+                    } else if (items['type'] === 'object') {
+                        // we are sure that there is a new object introduced here
+                        if (items.hasOwnProperty('properties') && items['properties'].hasOwnProperty('type')) {
+                            if (items['properties']['type'].hasOwnProperty('enum')) {
+                                _.forEach(items['properties']['type']['enum'], (name:string) => {
+                                    result.push(new Definition(name.toLowerCase(), Metaschema.cleanupDataschema(items, json), Metaschema.retrieveAcceptedElements(items, definitionsNames)));
+                                });
                             }
                         }
                     }
                 }
             }
+
             return result;
         }
 
-        private static cleanupDataschema(dataschema:{}):{} {
-            var result = {};
+        private static cleanupDataschema(dataschema:{}, json:{}):{} {
```

Comment:
I think the method name is not the best. If I can tell correctly this method does not alter its first argument, but instead folds overs the data schema and merges certain properties into a single `result` object, `flattenDataSchema` or `resolveDataSchema` would be more descriptive.